### ISSUE TITLE
end2end_test flake: assertion failed: ctx == elem_ctx

### ISSUE THREAD

**Issue by nicolasnoble** [grpc/grpc#1722]
```
D0522 03:44:51.999410080    4816 test_config.c:52] test slowdown: machine=1.000000 build=1.000000 total=1.000000
[==========] Running 22 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 22 tests from End2endTest
[ RUN      ] End2endTest.SimpleRpc
D0522 03:44:52.600553122    4816 iomgr.c:114] Waiting for 1 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.SimpleRpc (701 ms)
[ RUN      ] End2endTest.MultipleRpcs
D0522 03:44:53.601085423    4816 iomgr.c:114] Waiting for 1 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.MultipleRpcs (1000 ms)
[ RUN      ] End2endTest.RpcDeadlineExpires
D0522 03:44:53.901175167    4816 iomgr.c:114] Waiting for 1 iomgr objects to be destroyed and executing final callbacks
E0522 03:44:54.001226886    4889 tcp_client_posix.c:149] socket error: 104
D0522 03:44:54.001324788    4816 iomgr.c:114] Waiting for 0 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.RpcDeadlineExpires (401 ms)
[ RUN      ] End2endTest.RpcLongDeadline
D0522 03:44:55.001823456    4816 iomgr.c:114] Waiting for 1 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.RpcLongDeadline (1000 ms)
[ RUN      ] End2endTest.EchoDeadline
D0522 03:44:56.102438013    4816 iomgr.c:114] Waiting for 2 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.EchoDeadline (1101 ms)
[ RUN      ] End2endTest.EchoDeadlineForNoDeadlineRpc
D0522 03:44:57.102525348    4816 iomgr.c:114] Waiting for 1 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.EchoDeadlineForNoDeadlineRpc (1000 ms)
[ RUN      ] End2endTest.UnimplementedRpc
D0522 03:44:58.003133342    4816 iomgr.c:114] Waiting for 1 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.UnimplementedRpc (900 ms)
[ RUN      ] End2endTest.RequestStreamOneRequest
D0522 03:44:59.017269357    4816 iomgr.c:114] Waiting for 1 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.RequestStreamOneRequest (1014 ms)
[ RUN      ] End2endTest.RequestStreamTwoRequests
D0522 03:45:00.031926688    4816 iomgr.c:114] Waiting for 1 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.RequestStreamTwoRequests (1015 ms)
[ RUN      ] End2endTest.ResponseStream
D0522 03:45:01.131882175    4816 iomgr.c:114] Waiting for 2 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.ResponseStream (1100 ms)
[ RUN      ] End2endTest.BidiStream
I0522 03:45:02.032465765    5419 end2end_test.cc:158] recv msg hello0
I0522 03:45:02.032661692    5419 end2end_test.cc:158] recv msg hello1
I0522 03:45:02.032809591    5419 end2end_test.cc:158] recv msg hello2
D0522 03:45:02.132284065    4816 iomgr.c:114] Waiting for 1 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.BidiStream (1001 ms)
[ RUN      ] End2endTest.DiffPackageServices
[       OK ] End2endTest.DiffPackageServices (800 ms)
[ RUN      ] End2endTest.BadCredentials
E0522 03:45:03.033322694    4816 json_token.c:106] Invalid json string 
E0522 03:45:03.033347137    4816 credentials.c:755] Invalid input for service account credentials creation
[       OK ] End2endTest.BadCredentials (401 ms)
[ RUN      ] End2endTest.ClientCancelsRpc
D0522 03:45:04.434311238    4816 iomgr.c:114] Waiting for 2 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.ClientCancelsRpc (1100 ms)
[ RUN      ] End2endTest.ServerCancelsRpc
D0522 03:45:05.534562148    4816 iomgr.c:114] Waiting for 1 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.ServerCancelsRpc (1101 ms)
[ RUN      ] End2endTest.ClientCancelsRequestStream
D0522 03:45:06.635109207    4816 iomgr.c:114] Waiting for 2 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.ClientCancelsRequestStream (1100 ms)
[ RUN      ] End2endTest.ClientCancelsResponseStream
D0522 03:45:07.735207902    4816 iomgr.c:114] Waiting for 2 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.ClientCancelsResponseStream (1101 ms)
[ RUN      ] End2endTest.ClientCancelsBidi
I0522 03:45:08.636017295    5881 end2end_test.cc:158] recv msg hello0
I0522 03:45:08.636551608    5881 end2end_test.cc:158] recv msg hello1
D0522 03:45:08.835890062    4816 iomgr.c:114] Waiting for 2 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.ClientCancelsBidi (1100 ms)
[ RUN      ] End2endTest.RpcMaxMessageSize
D0522 03:45:09.936381928    4816 iomgr.c:114] Waiting for 2 iomgr objects to be destroyed and executing final callbacks
[       OK ] End2endTest.RpcMaxMessageSize (1101 ms)
[ RUN      ] End2endTest.SetPerCallCredentials
E0522 03:45:10.736659154    5933 metadata.c:564] assertion failed: ctx == elem_ctx
FAILED: bins/dbg/end2end_test [ret=-6, pid=4816]
```

Seen only once so far.

**Comment by ctiller** [comment]
This is worrying.

**Comment by ctiller** [comment]
@jboeuf 
This also likely warrants a re-architecture of how we handle metadata in credentials.

**Comment by nicolasnoble** [comment]
Got that one 5 times on a 5000 run.

**Comment by jboeuf** [comment]
OK... looks like we're going to have to change something. Craig do you want
to make a proposal?

**Comment by ctiller** [comment]
I do... next week when I've got hands to write.

**Comment by ctiller** [comment]
Is this function implementable?:

```c
grpc_creds_are_current_credentials_still_current(creds, some_token_returned_with_metadata_callback)
```

**Comment by jboeuf** [comment]
what does current mean here? That it is cached and still valid?

**Comment by ctiller** [comment]
Yeah. Trying to determine if the current set of strings needs to be
refreshed or not.

My thought is to:
- remove mdctx from credentials
- pass back just a set of strings when a metadata query is made
- cache the strings as mdelems on the channel data of the auth filter
- perform a quick check before sending them out for a given request

**Comment by jboeuf** [comment]
2 thoughts.

1. Will that work with credentials attached to just a call as well?
2. why do you need a check before sending them out? The credentials has
already done its work at this point.

**Comment by ctiller** [comment]
So for (1), I think it works just fine, except that we don't cache on the
channel.
And agreed for (2), we just need to do the quick check on subsequent
requests.

**Comment by jboeuf** [comment]
OK. As long as we do 2. outside of the credentials object (on the auth
filter for example), I'm ok with it.

**Comment by ctiller** [comment]
The auth filter should ask the question for sure. I'm not certain whether
it should be the creds object or something it supplies that answers the
question though. Looking to you for advice.

**Comment by jboeuf** [comment]
Let's talk tomorrow.

**Comment by ctiller** [comment]
Assigning since it looks like you're fixing it :)

**Comment by ctiller** [comment]
Tentatively placing this on the 0.9.1 release - per-call credentials are likely currently broken due to this.
